research/fast-api/index.md

```diff
@@ -0,0 +1,246 @@
+- async support
+- performance
+
+### Considerations
+
+- migration cost: refactoring, creating the models for documentation, rewriting of endpoints
+- benefits:
+  - we could get rid of non-used endpoints
+  - we could refactor/improve the functionality
+  - documentation of endpoints => easier dashboard development
+- alternatives:
+  - Sanic, Tornado, Falcon - lacking the functionalities such as validation, autodocumentation
+  - Django - too heavy
+  - none offer significant advantages over FastAPI/Flask
+
+## Migration requirements
+
+### Rewriting endpoints
+
+- convert Flask routes to FastAPI - replace Flask routes (`@app.route`) with FastAPI routes (`@app.get()`).
+- update request parsing, query params
+- response handling
+  - Pydantic models are recommended for response validation (raw dictionary responses can be returned also)
+- revisit error handling, utilise built-in support
+
+### Implementing models
+
+This is not strictly required (for the GET requests) but significantly improves the migration's value and is highly recommended.
+
+Previous attempt to document the endpoints with Flask:
+
+- https://github.com/packit/packit-service/pull/2089/files
+- the biggest issues: lot of duplication and additional code, change in serialization
+
+#### Using Pydantic with our existing models
+
+- retain current SQLAlchemy models for database interactions and add separate Pydantic schemas
+  to handle API validation, serialization, and documentation
+- support for integration with ORMs: https://docs.pydantic.dev/latest/concepts/models/#arbitrary-class-instances
+
+```python
+from pydantic import BaseModel, ConfigDict
+
+class UserSchema(BaseModel):
+    id: int
+    name: str
+    email: str
+
+    model_config = ConfigDict(from_attributes=True)
+```
+
+- advantages:
+  - lower risk of breaking something: wouldn't need to touch the DB layer
+  - separation: DB layer decoupled from the API layer - control on what gets exposed via the API without altering the persistence logic
+  - flexibility: schema can be updated independently
+- disadvantages:
+  - duplication
+  - manual mapping
+
+#### Migrating to [SQLModel](https://sqlmodel.tiangolo.com/)
+
+- a single model that serves both as database model and your API schema
+- advantages:
+  - no duplication
+  - automatically validated and documented endpoints
+  - cleaner codebase
+- disadvantages:
+  - higher migration overhead: rewriting models, increased risk
+  - less mature: it is newer than the established combination of SQLAlchemy + Pydantic
+
+```python
+from sqlmodel import SQLModel, Field
+
+class User(SQLModel, table=True):
+    id: int = Field(default=None, primary_key=True)
+    name: str
+    email: str
+
+```
+
+## Migration approach
+
+### Gradual rewrite
+
+- run Flask and FastAPI together
+- pick a few endpoints (e.g. file by file) and migrate
+- redirect traffic to FastAPI for migrated endpoints
+- repeat until everything is in FastAPI
+- remove Flask completely
```

Comment:
IMO I think keeping the endpoints for Flask is better and then introducing a documented FastAPI under something like `/api/v1/` to make the API documented. Then we wouldn't need to redirect traffic but simply mark it as deprecated until it breaks (or we remove it)